src/grid.rs

```diff
@@ -0,0 +1,207 @@
+    }
+
+    // Gets neighbors of a given position and counts how many of them are alive cells.
+    pub fn count_alive_neighbors(&self, x: usize, y: usize) -> u8 {
+        let mut count = 0;
+
+        let neighbors = self.get_neighbors(x, y);
+
+        for neighbor in neighbors{
+            if self.cells[neighbor.0][neighbor.1] == Cell::Alive {
+                count += 1;
+            }
+        }
+        
+        count
+    }
+
+    // Get all neighbors of a given position (not the actual Cell velues, but their positions in the grid)
+    // Handles the case in which there might be a position with fewer neighbors. Like [0,0]
+    pub fn get_neighbors(&self, x:usize, y:usize) -> Vec<(usize, usize)>{
+        let mut neighbors: Vec<(usize, usize)> = Vec::new();
+
+        // Relative positions
+        let deltas = [
+            (-1, -1), (-1, 0), (-1, 1), 
+            (0, -1), /*(0,0),*/ (0, 1), 
+            (1, -1), (1, 0), (1, 1),
+        ];
+
+        for (dx, dy) in &deltas {
+            // I use isize because result can be negative.
+            let new_x = x as isize + dx;
+            let new_y = y as isize + dy;
+
+            // Only push neighbor to neighbors if it's position is valid
+            if new_x >= 0 && new_x < self.rows() as isize && 
+                new_y >= 0 && new_y < self.cols() as isize {
+                neighbors.push((new_x as usize, new_y as usize));
+            }
+        }
+
+        neighbors
+    }
+
+    // Calculates next generation and updates the grid with that result.
+    pub fn update(&mut self) -> (){
+        self.cells = self.calculate_next_gen();
+    }
+
+    // For each cell in the grid calculates if in the next gen is going to be alive or not. Returns the next generation.
+    pub fn calculate_next_gen(&self) -> Vec<Vec<Cell>> {
+        let mut next_gen: Vec<Vec<Cell>> = vec![vec![Cell::Dead; self.cols()]; self.rows()];
+
+        for i in 0..self.rows(){
+            for j in 0..self.cols(){
+                let current_gen_cell = self.cells[i][j];
+                let alive_neighbors = self.count_alive_neighbors(i, j);
+                let next_gen_cell = match current_gen_cell {
+                    Cell::Alive if alive_neighbors == 2 || alive_neighbors == 3 => Cell::Alive,
+                    Cell::Dead if alive_neighbors == 3 => Cell::Alive,
+                    _ => Cell::Dead,
+                };
+                next_gen[i][j] = next_gen_cell;
+            }
+        }
+
+        next_gen
+    }
+}
+
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn create_preloaded_grid() -> Grid {
+        Grid::new_from_cells(vec![
+            vec![Cell::Dead, Cell::Alive, Cell::Dead, Cell::Dead],
+            vec![Cell::Alive, Cell::Alive, Cell::Dead, Cell::Dead],
+            vec![Cell::Dead, Cell::Dead, Cell::Alive, Cell::Dead],
+            vec![Cell::Dead, Cell::Dead, Cell::Dead, Cell::Alive],
+        ])
+    }
+
+    #[test]
+    fn test_grid_initialization() {
```

Comment:
The `test_` prefix isn't needed, I think that with the #[test] attribute is enough.

Reply:
Done!